driver/src/main/scala/org/mongodb/scala/internal/FlatMapObservable.scala

```diff
@@ -0,0 +1,122 @@
+        @volatile
+        private var started = false
+        @volatile
+        private var demand: Long = 0
+        @volatile
+        private var onCompleteCalled: Boolean = false
+
+        override def onSubscribe(subscription: Subscription): Unit = {
+
+          val masterSub = new Subscription() {
+            override def isUnsubscribed: Boolean = subscription.isUnsubscribed
+
+            override def request(n: Long): Unit = {
+              if (n < 1) {
+                throw new IllegalArgumentException(s"Number requested cannot be negative: $n")
+              }
+
+              val requestFirst = !started
+              val localDemand = addDemand(n)
+              if (!started) started = true
+
+              requestFirst match {
+                case true => subscription.request(1)
+                case false => nestedSubscription match {
+                  case Some(nestedSub) => nestedSub.request(localDemand)
+                  case None            => subscription.request(1)
+                }
+              }
+            }
+
+            override def unsubscribe(): Unit = subscription.unsubscribe()
+          }
+
+          outerSubscription = Some(masterSub)
+          observer.onSubscribe(masterSub)
+        }
+
+        override def onComplete(): Unit = {
+          onCompleteCalled = true
+          if (nestedSubscription.isEmpty) observer.onComplete()
+        }
+
+        override def onError(throwable: Throwable): Unit = observer.onError(throwable)
+
+        override def onNext(tResult: T): Unit = {
+          f(tResult).subscribe(
+            new Observer[S]() {
+              override def onError(throwable: Throwable): Unit = {
+                nestedSubscription = None
+                observer.onError(throwable)
+              }
+
+              override def onSubscribe(subscription: Subscription): Unit = {
+                nestedSubscription = Some(subscription)
+                subscription.request(demand)
+              }
+
+              override def onComplete(): Unit = {
+                nestedSubscription = None
+                onCompleteCalled match {
+                  case true => observer.onComplete()
+                  case false if demand > 0 =>
+                    addDemand(-1) // reduce demand by 1 as it will be incremented by the outerSubscription
+                    outerSubscription.get.request(1)
+                  case false => // No more demand
+                }
+              }
+
+              override def onNext(tResult: S): Unit = {
+                addDemand(-1)
+                observer.onNext(tResult)
+              }
+            }
+          )
+        }
+
+        /**
+         * Adds extra demand and protects against Longs rolling over
+         *
+         * @param extraDemand the amount of extra demand
+         * @return the updated demand
+         */
+        private def addDemand(extraDemand: Long): Long = {
```

Comment:
In that case volatile is not sufficient.  You need synchronization in this case.